language/move-vm/runtime/src/loader.rs

```diff
@@ -386,10 +371,105 @@ impl Loader {
         Ok((main, type_params))
     }
 
+    // The process of deserialization and verification is not and it must not be under lock.
+    // So when publishing modules through the dependency DAG it may happen that a different
+    // thread had loaded the module after this process fetched it from storage.
+    // Caching will take care of that by asking for each dependency module again under lock.
+    fn deserialize_and_verify_script(
+        &self,
+        script: &[u8],
+        data_store: &mut dyn DataStore,
+    ) -> VMResult<CompiledScript> {
+        let script = match CompiledScript::deserialize(script) {
+            Ok(script) => script,
+            Err(err) => {
+                error!("[VM] deserializer for script returned error: {:?}", err);
+                let error = vm_status(Location::default(), StatusCode::CODE_DESERIALIZATION_ERROR)
+                    .append(err);
+                return Err(error);
+            }
+        };
+
+        match self.verify_script(&script) {
+            Ok(_) => {
+                // verify dependencies
+                let deps = load_script_dependencies(&script);
+                let mut dependencies = vec![];
+                for dep in &deps {
+                    dependencies.push(self.load_module(dep, data_store)?);
+                }
+                self.verify_script_dependencies(script, dependencies)
+            }
+            Err(err) => {
+                error!(
+                    "[VM] bytecode verifier returned errors for script: {:?}",
+                    err
+                );
+                Err(err)
+            }
+        }
+    }
+
+    // Script verification steps.
+    // See `verify_module()` for module verification steps.
+    fn verify_script(&self, script: &CompiledScript) -> VMResult<()> {
+        DuplicationChecker::verify_script(&script)?;
+        SignatureChecker::verify_script(&script)?;
+        InstructionConsistency::verify_script(&script)?;
+        constants::verify_script(&script)?;
+        CodeUnitVerifier::verify_script(&script)?;
+        verify_main_signature(&script)
+    }
```

Comment:
This looks like the same code in verifier.rs? Why the duplication?

Reply:
I want the runtime to have its own concept of verification and I do not want to use a common function.
Both `verify_module` and `verify_script` are implemented by the runtime.
Maybe I could point that out in the comment but it really seem too much information.
Bottom line I do not want the runtime to share anything with anybody when it comes to core logic.
Of course that has a lot of not so nice side effects together with a lot of very nice, and truth is, I don't know what's right.
But I really do not want to share for now... I want to be really selfish here... ;) 
thoughts?